### ISSUE TITLE
Estimate dependence probability from pairwise columns does not run in parallel

### ISSUE THREAD

**Issue by marcoct** [probcomp/bayeslite#257]
There are 32 models in the bdb. Also, there are 32 cpus allocated:
```
cc = ccme.MultiprocessingEngine(seed=seed, cpu_count=32)
```
However, only one cpu is being used.

Is this supposed to be the case? What is parallelized across models and what is not?

**Comment by marcoct** [comment]
It turns out that for 6000 rows and 1200 columns this doesn't take too long (5 minutes), so this is acceptable performance for this particular dataset.

**Comment by axch** [comment]
The answer to the actual question is that only analysis is parallelized.  Parallelizing queries is at least a little tricky, because the overhead of multiprocessing in Python is pretty high, so it would be necessary to choose good grain sizes at which to do it.  Also, the overhead increases further in our case because there are in-query caches, the work of warming which a naive parallelization would not be able to share across processes.

In light of that, should we close the issue?  Move it to the nice to have / someday milestone?

**Comment by riastradh-probcomp** [comment]
In principle, it is easy to imagine splitting a simple query over as many parts of the table as there are cores, or to `shard' it, to use the word in vogue -- then the overhead of communication is limited and even if we didn't have inter-CPU communication for the caches, each CPU would individually be able to take advantage of a cache of its own.

But the query engine we use is sqlite3's, and to my knowledge there is no way to make sqlite3 take advantage of multiple cores when executing a single query.  It is no good to just dispatch every bql_column_dependence_probability call to a multiprocessing pool, because each one has to return synchronously before the next one starts.

It is conceivable that for limited classes of queries, we could shard explicitly at the top level, e.g. we could turn

    ESTIMATE ... FROM t

into an explicitly evaluated union of

    ESTIMATE ... FROM t LIMIT 100 OFFSET 0
    ESTIMATE ... FROM t LIMIT 100 OFFSET 100
    ESTIMATE ... FROM t LIMIT 100 OFFSET 200

Handling `WHERE`, `ORDER BY`, and `GROUP BY` might be doable similarly with a little more work.

**Comment by riastradh-probcomp** [comment]
Another issue is that Python multiprocessing pools break ^C, which is sort of OKish if you're using them only ANALYZE and not trying to get work done interactively, but definitely not OK if you're interactively trying to explore your data.  So we really need to find or write a better replacement for Python multiprocessing.

**Comment by gregory-marton** [comment]
I'm inclined to close this as working-as-intended, after adding an appropriate maintainer note. Where in the code is a good place to refer to the discussion in this thread?

**Comment by riastradh-probcomp** [comment]
We ought to have an open issue for parallelizing model estimators.  It's a nontrivial engineering problem that we're not going to address in the foreseeable future, but nevertheless worth documenting for future engineering and to answer the oft-asked question of why an embarrassingly parallelizable slow query like `ESTIMATE MUTUAL INFORMATION FROM PAIRWISE COLUMNS OF t` is taking forever and spinning only a single CPU.  This issue seems as good a place as any for the purpose.
